animation_nodes/nodes/matrix/distribute_matrices.pyx

```diff
@@ -385,6 +411,43 @@ class DistributeMatricesNode(bpy.types.Node, AnimationNode):
         count = len(spline.points)
         return spline.getDistributedMatrices(count, 0, 1, "RESOLUTION")
 
+    def execute_HexagonalGrid(self, int xDivisions, int yDivisions, float xSize, float ySize):
+        cdef float xDistance = xSize * 1.5
+        cdef float yDistance = ySize * sqrt(3)
+
+        if self.hexagonalDistanceMode == "SIZE":
+            xDistance /= max(xDivisions - 1, 1)
+            yDistance /= max(yDivisions - 1, 1)
+
+        cdef float xOffset = 0
+        cdef float yOffset = 0
+
+        if self.hexGridCenterX:
+            xOffset = xDistance * (xDivisions - 1) / 2
+        if self.hexGridCenterY:
+            yOffset = yDistance * (yDivisions - 1) / 2
+
+        cdef Vector3 translation = toVector3((0, 0, 0))
+        cdef Matrix4x4List matrices = Matrix4x4List(length = xDivisions * yDivisions)
+        cdef Py_ssize_t i, j, index
+        cdef float x, y
+
+        for i in range(xDivisions):
+            x = i * xDistance
+
+            for j in range(yDivisions):
+                if (i % 2) == 0:
+                    y = j * yDistance
+                else:
+                    y = (j + 0.5) * yDistance
```

Comment:
The for loops should be switched. The outer loop should be the one for `yDivisions` and the inner loop should be for `xDivisions`.

There are two reasons why you should do this:

1- y is more expensive to compute and is conditional, so it would make sense to compute it once for every x in the same row.
2. You write the matrices in a row major order. When you have `yDivisions` as the inner loop, you might jump to a different cache line for every single iteration, which means bad cache coherency. While if you have `xDivisions` as the inner loop, you always write matrices sequentially, so you get ideal spacial cache coherency.

While at it, try computing y as follows `y = (j + 0.5 * (i % 2)) * yDistance`. It might give better performance for high number of matrices. Try it maybe.

Reply:
> The for loops should be switched. The outer loop should be the one for `yDivisions` and the inner loop should be for `xDivisions`.
> 
> There are two reasons why you should do this:
> 
> 1- y is more expensive to compute and is conditional, so it would make sense to compute it once for every x in the same row.
> 2. You write the matrices in a row major order. When you have `yDivisions` as the inner loop, you might jump to a different cache line for every single iteration, which means bad cache coherency. While if you have `xDivisions` as the inner loop, you always write matrices sequentially, so you get ideal spacial cache coherency.
> 
> While at it, try computing y as follows `y = (j + 0.5 * (i % 2)) * yDistance`. It might give better performance for high number of matrices. Try it maybe.

Can u provide a pseudocode for this?

Reply:
```python
for j in range(yDivisions):
    for x in range(xDivisions):
        x = i * xDistance
        y = (j + 0.5 * (i % 2)) * yDistance
```

Reply:
@OmarEmaraDev Updated